# -*- mode: org; mode: org-indent; mode: visual-line -*-

#+TITLE: Project README
#+DATE: 2016-09-30 Fri
#+AUTHOR: Becky Stodola, Mark Neider, Nick Daly

* Prerequisites

1. Install [[https://cygwin.com/][cygwin]].
2. Install:
   - gzip
   - python3
   - sqlite3
   - wget
3. Install [[http://sqlitebrowser.org/][SQLite DB Browser]].
4. Install [[https://tortoisegit.org/][Tortoise Git]].
5. Install [[http://plantuml.com/][PlantUml]] to [[src/createDiagrams.sh][convert schemas to diagrams]].

Which things do you need?  Run ~which (thing)~ to find out.

* Downloading the Database

Open up a command window.  In the directory with this README file, run:

: cd genome-explorer
: sh src/createDb.sh

Or:

: bash src/createDb.sh

That'll  download the database into the ~bin~ directory.

* Checkpoints

** DONE Checkpoint 1 - https://docs.google.com/document/d/11Z1eH0OzM--XwBi3L11mag8DigBeM1Hh_-UeI8enRmA/edit?usp=sharing

We've created the overview document.

** TODO Checkpoint 2 (due 10/13) - https://docs.google.com/document/d/19jVVghHHxMClzC_J8lViPCmsw4nvFJ-cPNXOxt6plcU/edit?usp=sharing

We need to identify tables and create the ER diagram:

- [ ] Tables identified.
- [ ] ER diagram created.

** TODO Checkpoint 3 (due 10/27) - https://docs.google.com/document/d/1dYJqh1c9UnNLvgrt5KcjDFYkjHULCN4H0hjT049A4qs/edit?usp=sharing

** TODO Checkpoint 4 (due 11/17) - https://docs.google.com/document/d/11SSkbGrPP14u647pHf5v-dFdvDMoR8HHPd5Uqig1cSY/edit?usp=sharing

* Outline of Schema and Tables

NCBI table/column data dictionary:  https://www.ncbi.nlm.nih.gov/SNP/snp_db_list_table.cgi

** Schemas (also an index and constraints file in each directory)

- Shared schema ftp://ftp.ncbi.nih.gov/snp/database/shared_schema/dbSNP_main_table.sql.gz
- Human schema ftp://ftp.ncbi.nih.gov/snp/database/organism_schema/human_9606/human_9606_table.sql.gz

** Tables

- All shared tables?   ftp://ftp.ncbi.nih.gov/snp/database/shared_data/

- Basic SNP data
  - SNP - 1.8 GB
  - SNPSubSNPLink - 4.2 GB
  - SubSNP - 7.5 GB
  - Synonym - type and name from subsnp_id

- Frequency info (lots of different tables with different slices)
  - Allele - Unique alleles (single nucleotide A, T, C, or G) for a given location in DNA
  - AlleleFreqBySsPop - Frequency of a given allele (single nucleotide) at a given submitted SNP (SS#) by population
    - subsnp_id - foreign key to SubSNP table
    - pop_id - foreign key to Population table
    - allele_id - foreign key to Allele table
  - GtyFreqBySsPop: This table saves computed genotype frequency per subsnp and population.
    - UniGty required
  - SNPAlleleFreq?: This table stores the average allele frequency from snp_id and allele_id. 1.2GB

- Clinical significance and other hopefully interesting info
  - SNPClinSig: Clinical significance of a SNP
  - ClinSigCode: Clinical significance numeric codes and category titles (Benign, Pathogenic, other, etc)
  - SNPPubmed: Links to PubMed ID for published literature on a SNP

- Population data
  - Population
  - PopLine: Population description text is stored here line by line
  - UniGty: Genotype
