# -*- mode: org; mode: org-indent; mode: visual-line -*-

#+TITLE: Project README
#+DATE: 2016-09-30 Fri
#+AUTHOR: Becky Stodola, Mark Neider, Nick Daly

* Prerequisites

1. Install [[https://cygwin.com/][cygwin]].
2. Install:
   - gzip
   - python3
   - sqlite3
   - wget
3. Install [[http://sqlitebrowser.org/][SQLite DB Browser]].
4. Install [[https://tortoisegit.org/][Tortoise Git]].
5. Install [[http://plantuml.com/][PlantUml]] to [[src/createDiagrams.sh][convert schemas to diagrams]].

Which things do you need?  Run ~which (thing)~ to find out.

* Downloading the Database

Open up a command window.  In the directory with this README file, run:

: cd genome-explorer
: sh src/createDb.sh

Or:

: bash src/createDb.sh

That'll  download the database into the ~bin~ directory.

* Updating the ER Diagram

Open a command window.  In the directory with this README file, run the following.  This will rebuild the .png file in the bin directory.

: sh src/createErDiagram.sh

* Checkpoints

** DONE Checkpoint 1 - Grade 30/30

https://docs.google.com/document/d/11Z1eH0OzM--XwBi3L11mag8DigBeM1Hh_-UeI8enRmA/edit?usp=sharing

We've created the overview document.

Feedback from professor:  This is a well-written proposal. The dataset is very well-chosen and tasks are focused enough so that we don't get lost in the size of data. It is very good that you are aware of the gaps in the data which may create challenges in the assessment. For the assessment, I would recommend we define some criteria where the "interesting" traits are defined (my understanding from the write-up: interesting = common trait" while in my thought, interesting = uncommon trait"). Also a set of scenarios with inputs and expected outputs would help us in evaluation.

** TODO Checkpoint 2 (due 10/13)

https://docs.google.com/document/d/19jVVghHHxMClzC_J8lViPCmsw4nvFJ-cPNXOxt6plcU/edit?usp=sharing

We need to identify tables and create the ER diagram:

- [X] Tables identified.
- [X] ER diagram created.
- [ ] Finish documentation of features, functionality, queries, and UI.
- [ ] Add more constraints to ER diagram and description.

** TODO Checkpoint 3 (due 10/27)

https://docs.google.com/document/d/1dYJqh1c9UnNLvgrt5KcjDFYkjHULCN4H0hjT049A4qs/edit?usp=sharing

** TODO Checkpoint 4 (due 11/17)

https://docs.google.com/document/d/11SSkbGrPP14u647pHf5v-dFdvDMoR8HHPd5Uqig1cSY/edit?usp=sharing

* Outline of Schema and Tables

NCBI table/column data dictionary:  https://www.ncbi.nlm.nih.gov/SNP/snp_db_list_table.cgi

** Schemas (also an index and constraints file in each directory)

- Shared schema ftp://ftp.ncbi.nih.gov/snp/database/shared_schema/dbSNP_main_table.sql.gz
- Human schema ftp://ftp.ncbi.nih.gov/snp/database/organism_schema/human_9606/human_9606_table.sql.gz

** Tables

- Shared tables   ftp://ftp.ncbi.nih.gov/snp/database/shared_data/
 - Allele - Unique alleles (single nucleotide A (6), T (4), C (7), or G (2)) for a given location in DNA
 - UniGty - Genotypes, including both alleles from the two DNA strands at a given location (CC, TA, etc).
   - unigty_id - Assigned ID for the genotype, integer
   - gty_str - String representation of the genotype (example:  C/G), text
   - allele_id_1 - ID of the first allele of the genotype, integer
   - allele_id_2 - ID of the second allele of the genotype, integer
   - create_time - Time when the record was created, datetime

- Human tables
 - Basic SNP data
   - SNP - 1.8 GB
    - snp_id - Primary key, identifier for the SNP location
   - SNPSubSNPLink - 4.2 GB
   - SubSNP - Data for a submitted instance of a SNP.  7.5 GB
     - subsnp_id - Primary key, the unique ID assigned to a submitted value for a SNP (referred to as ss# in NCBI documentation)
   - Synonym - type and name from subsnp_id

 - Frequency info (lots of different tables with different slices)
   - SNPAlleleFreq: This table stores the average allele frequency from snp_id and allele_id. 1.2GB
     - snp_id  - Primary key, identifier for the SNP location, integer
     - allele_id - Foreign key to Allele table, integer
     - freq - Frequency of this allele for this SNP, float
   - SNPGtyFreq - Average genotype frequency for a refSNP(rs#). 249 MB
     - snp_id - rs#, int
     - unigty_id - Foreign key to UniGty table, int
     - ind_cnt - Count of individuals with the genotype specified in unigty_id, float
     - freq - Frequency of this genotype, float
     - last_updated_time - datetime
   - AlleleFreqBySsPop - Frequency of a given allele (single nucleotide) at a given submitted SNP (SS#) by population
      - subsnp_id - foreign key to SubSNP table
      - pop_id - foreign key to Population table
      - allele_id - foreign key to Allele table
   - GtyFreqBySsPop: This table saves computed genotype frequency per subsnp and population.
      - subsnp_id
      - pop_id
      - unigty_id - UniGty table required
      - freq

 - Clinical significance and other hopefully interesting info
   - SNPClinSig: Clinical significance of a SNP
   - ClinSigCode: Clinical significance numeric codes and category titles (Benign, Pathogenic, other, etc)
   - SNPPubmed: Links to PubMed ID for published literature on a SNP

 - Population data
   - Population
   - PopLine: Population description text is stored here line by line
